[Merged by Bors] - feat(Algebra/Polynomial): small and useful lemmas

SnirBroshi opened the pull request:
---
Moves these 4 theorems from `Mathlib.Algebra.Polynomial.Monic` to `Mathlib.Algebra.Polynomial.Degree.Lemmas`:
- `degree_map_eq_of_injective`
- `natDegree_map_eq_of_injective`
- `leadingCoeff_map_of_injective`
- `nextCoeff_map`

`Monic.lean` imports `Lemmas.lean` so users are not affected.

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

kckennylau commented on Mathlib/Algebra/Polynomial/Degree/Lemmas.lean:
```diff
@@ -272,6 +272,19 @@ theorem natDegree_map_eq_iff {f : R →+* S} {p : Polynomial R} :
   · simp_rw [h, ne_eq, or_true, iff_true, ← Nat.le_zero, ← h, natDegree_map_le]
   simp_all [natDegree, WithBot.unbotD_eq_unbotD_iff]
 
+theorem degree_map_eq_of_injective {f : R →+* S} {p : Polynomial R} (hf : Function.Injective f) :
+    (p.map f).degree = p.degree := by
+  simp [hf, map_ne_zero_iff, ne_or_eq]
+
```
[Polynomial.degree_map_eq_of_injective](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Polynomial/Monic.html#Polynomial.degree_map_eq_of_injective)

kckennylau commented on Mathlib/Algebra/Polynomial/Degree/Lemmas.lean:
```diff
@@ -324,6 +337,13 @@ lemma natDegree_eq_one : p.natDegree = 1 ↔ ∃ a ≠ 0, ∃ b, C a * X + C b =
   · rintro ⟨a, ha, b, rfl⟩
     simp [ha]
 
+theorem isRoot_subsingleton_of_degree_eq_one [IsLeftCancelMul R] [IsRightCancelAdd R]
```
```suggestion
theorem subsingleton_isRoot_of_degree_eq_one [IsLeftCancelMul R] [IsRightCancelAdd R]
```

kckennylau commented on Mathlib/Algebra/Polynomial/Eval/Coeff.lean:
```diff
@@ -61,6 +61,12 @@ theorem coeff_zero_eq_eval_zero (p : R[X]) : coeff p 0 = p.eval 0 :=
 theorem zero_isRoot_of_coeff_zero_eq_zero {p : R[X]} (hp : p.coeff 0 = 0) : IsRoot p 0 := by
   rwa [coeff_zero_eq_eval_zero] at hp
 
+theorem coeff_zero_eq_zero_of_zero_isRoot {p : R[X]} (hp : IsRoot p 0) : p.coeff 0 = 0 := by
+  rwa [coeff_zero_eq_eval_zero]
+
+theorem zero_isRoot_iff_coeff_zero_eq_zero {p : R[X]} : IsRoot p 0 ↔ p.coeff 0 = 0 :=
+  ⟨coeff_zero_eq_zero_of_zero_isRoot, zero_isRoot_of_coeff_zero_eq_zero⟩
```
You could probably combine all these lemmas:

```suggestion
theorem zero_isRoot_iff_coeff_zero_eq_zero {p : R[X]} : IsRoot p 0 ↔ p.coeff 0 = 0 := by
  rw [coeff_zero_eq_eval_zero, IsRoot]

alias ⟨coeff_zero_eq_zero_of_zero_isRoot, zero_isRoot_of_coeff_zero_eq_zero⟩ :=
  zero_isRoot_iff_coeff_zero_eq_zero
```

kckennylau commented on Mathlib/Algebra/Polynomial/Eval/Coeff.lean:
```diff
@@ -223,4 +229,17 @@ end Map
+variable [DivisionRing R] [Nontrivial S] [Semiring S]
+variable {f : R →+* S} {p : R[X]}
+
+theorem map_eq_zero_iff_from_divisionRing : p.map f = 0 ↔ p = 0 :=
+  Polynomial.map_eq_zero_iff <| RingHom.injective_from_divisionRing f
+
+theorem map_ne_zero_iff_from_divisionRing : p.map f ≠ 0 ↔ p ≠ 0 :=
+  Polynomial.map_ne_zero_iff <| RingHom.injective_from_divisionRing f
```
Please generalise to the conditions of [RingHom.injective](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/SimpleRing/Basic.html#RingHom.injective).

kckennylau commented on Mathlib/Algebra/Polynomial/Eval/Defs.lean:
```diff
@@ -546,9 +546,14 @@ protected theorem map_list_prod (L : List R[X]) : L.prod.map f = (L.map <| map f
 theorem eval_map (x : S) : (p.map f).eval x = p.eval₂ f x :=
   (eval₂_eq_eval_map f).symm
 
+lemma eval_map' (x : R) :
```
```suggestion
lemma eval_map_apply (x : R) :
```

kckennylau commented on Mathlib/Algebra/Polynomial/Eval/Degree.lean:
```diff
@@ -180,15 +177,30 @@ theorem natDegree_map_of_leadingCoeff_ne_zero (f : R →+* S) (hf : f (leadingCo
+    h ▸ RingHom.isUnit_map _ hf
+
+variable (f) in
+theorem natDegree_map_of_leadingCoeff_isUnit [Nontrivial S] (hf : IsUnit <| p.leadingCoeff) :
+    (p.map f).natDegree = p.natDegree :=
+  natDegree_eq_natDegree <| degree_map_of_leadingCoeff_isUnit f hf
```
```suggestion
theorem natDegree_map_of_isUnit_leadingCoeff [Nontrivial S] (hp : IsUnit p.leadingCoeff) :
    (p.map f).natDegree = p.natDegree :=
  natDegree_eq_natDegree <| degree_map_of_leadingCoeff_isUnit f hp
```

kckennylau commented on Mathlib/Algebra/Polynomial/Eval/Degree.lean:
```diff
@@ -180,15 +177,30 @@ theorem natDegree_map_of_leadingCoeff_ne_zero (f : R →+* S) (hf : f (leadingCo
+  natDegree_eq_natDegree <| degree_map_of_leadingCoeff_isUnit f hf
+
+variable (f) in
+theorem leadingCoeff_map_of_leadingCoeff_isUnit [Nontrivial S] (hf : IsUnit <| p.leadingCoeff) :
+    (p.map f).leadingCoeff = f (p.leadingCoeff) :=
+  leadingCoeff_map_of_leadingCoeff_ne_zero _ <| fun h ↦ not_isUnit_zero <|
+    h ▸ RingHom.isUnit_map _ hf
```
```suggestion
theorem leadingCoeff_map_of_isUnit_leadingCoeff [Nontrivial S] (hp : IsUnit p.leadingCoeff) :
    (p.map f).leadingCoeff = f (p.leadingCoeff) :=
  leadingCoeff_map_of_leadingCoeff_ne_zero _ (f.isUnit_map hp).ne_zero
```

kckennylau commented on Mathlib/Algebra/Polynomial/Eval/Degree.lean:
```diff
@@ -180,15 +177,30 @@ theorem natDegree_map_of_leadingCoeff_ne_zero (f : R →+* S) (hf : f (leadingCo
+theorem natDegree_map_of_leadingCoeff_isUnit [Nontrivial S] (hf : IsUnit <| p.leadingCoeff) :
+    (p.map f).natDegree = p.natDegree :=
+  natDegree_eq_natDegree <| degree_map_of_leadingCoeff_isUnit f hf
+
+variable (f) in
+theorem leadingCoeff_map_of_leadingCoeff_isUnit [Nontrivial S] (hf : IsUnit <| p.leadingCoeff) :
+    (p.map f).leadingCoeff = f (p.leadingCoeff) :=
+  leadingCoeff_map_of_leadingCoeff_ne_zero _ <| fun h ↦ not_isUnit_zero <|
+    h ▸ RingHom.isUnit_map _ hf
 
 end Map
 
 end Semiring
 
-section CommSemiring
-
-section Eval
+section CommSemiring.Eval
```
This is not a namespace, I don't think this change is necessary.

kckennylau commented on Mathlib/Algebra/Polynomial/Div.lean:
```diff
@@ -704,6 +705,11 @@ lemma le_rootMultiplicity_mul {p q : R[X]} (x : R) (hpq : p * q ≠ 0) :
   rw [le_rootMultiplicity_iff hpq, pow_add]
   gcongr <;> apply pow_rootMultiplicity_dvd
 
+lemma le_rootMultiplicity_div {p q : R[X]} (hq : q ≠ 0) (hpq : p ∣ q) (x : R) :
```
```suggestion
lemma rootMultiplicity_le_rootMultiplicity_of_dvd {p q : R[X]} (hq : q ≠ 0) (hpq : p ∣ q) (x : R) :
```

kckennylau commented on Mathlib/Algebra/Polynomial/Degree/Lemmas.lean:
```diff
@@ -324,6 +343,13 @@ lemma natDegree_eq_one : p.natDegree = 1 ↔ ∃ a ≠ 0, ∃ b, C a * X + C b =
+theorem subsingleton_isRoot_of_natDegree_eq_one [IsLeftCancelMulZero R] [IsRightCancelAdd R]
+    (h : p.natDegree = 1) : { x | IsRoot p x }.Subsingleton := by
+  intro x hx y hy
+  have ⟨a, ha, b, hp⟩ := Polynomial.natDegree_eq_one.mp h
+  simp [← hp] at hx hy
```
```suggestion
  obtain ⟨a, ha, b, rfl⟩ := Polynomial.natDegree_eq_one.mp h
  simp only [IsRoot.def, eval_add, eval_mul_X, eval_C, Set.mem_setOf_eq] at hx hy
```
You have a non-terminal simp, i.e. your simp does not close a goal, so you should convert that to `simp?`, which we call "squeezing the simp"; and then when you click on the result, you will get the line I have suggested.

kckennylau commented on Mathlib/Algebra/Polynomial/Roots.lean:
```diff
@@ -111,6 +111,23 @@ theorem ne_zero_of_mem_roots (h : a ∈ p.roots) : p ≠ 0 :=
+theorem roots_zero_iff_isRoot_bot (hp0 : p ≠ 0) : p.roots = 0 ↔ p.IsRoot = ⊥ := by
+  constructor <;> intro h
+  · ext a
+    simp
```
non-terminal simp, please squeeze simp (i.e. put `simp?` and then click the result)

kckennylau commented on Mathlib/Algebra/Polynomial/Roots.lean:
```diff
@@ -710,12 +727,32 @@ theorem Monic.isUnit_leadingCoeff_of_dvd {a p : R[X]} (hp : Monic p) (hap : a 
     IsUnit a.leadingCoeff :=
   isUnit_of_dvd_one (by simpa only [hp.leadingCoeff] using leadingCoeff_dvd_leadingCoeff hap)
 
+theorem Irreducible.roots_card_le_one (hirr : Irreducible p) : p.roots.card ≤ 1 := by
+  by_contra hcard
+  have ⟨q, hq⟩ := p.prod_multiset_X_sub_C_dvd
+  have ⟨x₁, hx₁⟩ := card_pos_iff_exists_mem.mp <| show 0 < p.roots.card by cutsat
+  have ⟨q₁, hq₁⟩ := exists_cons_of_mem hx₁
+  have := hq₁ ▸ Multiset.card_cons x₁ q₁
+  have ⟨x₂, hx₂⟩ := card_pos_iff_exists_mem.mp <| show 0 < q₁.card by cutsat
+  have ⟨q₂, hq₂⟩ := exists_cons_of_mem hx₂
+  simp [hq₁, hq₂, mul_assoc] at hq
+  obtain hu | hu := hirr.isUnit_or_isUnit hq <;> apply degree_eq_zero_of_isUnit at hu <;> simp at hu
+  have := Nat.WithBot.add_eq_zero_iff.mp hu |>.left
+  contradiction
```
```suggestion
theorem roots_congr {p q : R[X]} (h : Associated p q) : p.roots = q.roots := by
  obtain ⟨u, rfl⟩ := h
  rw [eq_C_of_degree_eq_zero <| degree_coe_units u, mul_comm,
    roots_C_mul _ (coeff_coe_units_zero_ne_zero u)]

theorem Irreducible.roots_card_le_one (hirr : Irreducible p) : p.roots.card ≤ 1 := by
  by_contra hcard
  apply hcard
  have ⟨x, hx⟩ := card_pos_iff_exists_mem.mp <| show 0 < p.roots.card by cutsat
  rw [← roots_congr <| (irreducible_X_sub_C _).associated_of_dvd hirr <| dvd_iff_isRoot.mpr <|
    isRoot_of_mem_roots hx]
```

`roots_congr` can probably be moved right after `roots_C_mul`.

kckennylau replied:
```suggestion
theorem Irreducible.roots_card_le_one (hirr : Irreducible p) : p.roots.card ≤ 1 := by
  obtain hp | ⟨x, hx⟩ := p.roots.empty_or_exists_mem
  · simp [hp]
  convert p.card_roots'
  exact (natDegree_eq_of_degree_eq_some <| degree_eq_one_of_irreducible_of_root hirr
    (isRoot_of_mem_roots hx)).symm
```

kckennylau commented on Mathlib/Algebra/Polynomial/Roots.lean:
```diff
@@ -785,6 +822,26 @@ theorem map_roots_le_of_injective [IsDomain A] [IsDomain B] (p : A[X]) {f : A 
   · simp only [hp0, roots_zero, Multiset.map_zero, Polynomial.map_zero, le_rfl]
   exact map_roots_le ((Polynomial.map_ne_zero_iff hf).mpr hp0)
 
+theorem map_roots_card_le_degree {A B : Type*} [Semiring A] [CommRing B] [IsDomain B]
+    {f : A →+* B} (p : A[X]) (h : p.map f ≠ 0) : (p.map f).roots.card ≤ p.degree :=
+  card_roots h |>.trans degree_map_le
```
`h` is not necessary

NotWearingPants replied:
Almost, generalized to `p ≠ 0`

kckennylau commented on Mathlib/Algebra/Polynomial/Roots.lean:
```diff
@@ -785,6 +822,26 @@ theorem map_roots_le_of_injective [IsDomain A] [IsDomain B] (p : A[X]) {f : A 
+    {f : A →+* B} (p : A[X]) : (p.map f).roots.card ≤ p.natDegree :=
+  card_roots' _ |>.trans natDegree_map_le
+
+open Classical in
+theorem map_roots_filter_range_eq_roots_map [IsDomain A] [IsDomain B]
+    {f : A →+* B} (hf : Function.Injective f) (p : A[X]) :
+    (p.map f).roots.filter (· ∈ f.range) = p.roots.map f := by
```
Please assume the `DecidablePred` instance instead of `open`ing `Classical`.

kckennylau replied:
This is to avoid "diamond", i.e. right now if someone provides another instance that the predicate is decidable, then they will not be able to use your theorem.

kckennylau commented on Mathlib/FieldTheory/IsAlgClosed/Basic.lean:
```diff
@@ -108,13 +108,52 @@ theorem roots_eq_zero_iff [IsAlgClosed k] {p : k[X]} :
+  have ⟨_, _, hdeg, hroots⟩ := exists_prod_multiset_X_sub_C_mul p
+  simp [← hdeg, roots_eq_zero_iff_natDegree_zero.mp hroots]
+
+theorem map_roots_card_eq_natDegree_of_leadingCoeff_ne_zero {A B : Type*} [CommRing A] [Field B]
+    [IsAlgClosed B] (f : A →+* B) {p : A[X]} (hf : f p.leadingCoeff ≠ 0) :
+    (p.map f).roots.card = p.natDegree :=
+  natDegree_map_of_leadingCoeff_ne_zero _ hf ▸ roots_card_eq_natDegree
+
+theorem map_roots_card_eq_natDegree_of_isUnit_leadingCoeff {A B : Type*} [CommRing A] [Field B]
+    [IsAlgClosed B] (f : A →+* B) {p : A[X]} (h : IsUnit p.leadingCoeff) :
+    (p.map f).roots.card = p.natDegree :=
+  natDegree_map_of_isUnit_leadingCoeff f h ▸ roots_card_eq_natDegree
```
does A have to be a comm ring?

NotWearingPants replied:
Good catch, generalized to `Semiring`

kckennylau commented on Mathlib/FieldTheory/IsAlgClosed/Basic.lean:
```diff
@@ -108,13 +108,52 @@ theorem roots_eq_zero_iff [IsAlgClosed k] {p : k[X]} :
+    [Field B] [IsAlgClosed B] (f : A →+* B) {p : A[X]} : (p.map f).roots.card = p.natDegree :=
+  natDegree_map_from_simpleRing f p ▸ roots_card_eq_natDegree
+
+theorem aroots_card_eq_natDegree_of_leadingCoeff_ne_zero {A B : Type*} [CommRing A] [Field B]
+    [IsAlgClosed B] [Algebra A B] {p : A[X]} (hf : algebraMap A B p.leadingCoeff ≠ 0) :
+    (p.aroots B).card = p.natDegree :=
+  map_roots_card_eq_natDegree_of_leadingCoeff_ne_zero _ hf
+
+theorem aroots_card_eq_natDegree_of_isUnit_leadingCoeff {A B : Type*} [CommRing A] [Field B]
+    [IsAlgClosed B] [Algebra A B] {p : A[X]} (h : IsUnit p.leadingCoeff) :
+    (p.aroots B).card = p.natDegree :=
+  map_roots_card_eq_natDegree_of_isUnit_leadingCoeff _ h
+
+theorem aroots_card_eq_natDegree_from_field {A B : Type*} [CommRing A] [IsSimpleRing A]
+    [Field B] [IsAlgClosed B] [Algebra A B] {p : A[X]} : (p.aroots B).card = p.natDegree :=
+  map_roots_card_eq_natDegree_from_simpleRing _
```
does A have to be a comm ring?

erdOne commented:
!bench

erdOne commented on Mathlib/Algebra/Polynomial/Roots.lean:
```diff
@@ -111,6 +111,23 @@ theorem ne_zero_of_mem_roots (h : a ∈ p.roots) : p ≠ 0 :=
 theorem isRoot_of_mem_roots (h : a ∈ p.roots) : IsRoot p a :=
   (mem_roots'.1 h).2
 
+theorem roots_zero_iff_isRoot_bot (hp0 : p ≠ 0) : p.roots = 0 ↔ p.IsRoot = ⊥ := by
```
```suggestion
theorem roots_eq_zero_iff_isRoot_eq_bot (hp0 : p ≠ 0) : p.roots = 0 ↔ p.IsRoot = ⊥ := by
```

erdOne commented on Mathlib/Algebra/Polynomial/Roots.lean:
```diff
@@ -111,6 +111,23 @@ theorem ne_zero_of_mem_roots (h : a ∈ p.roots) : p ≠ 0 :=
+    exact mem_roots hp0 |>.not.mp <| by simp [h]
+  · exact eq_zero_of_forall_notMem fun x hx ↦ h ▸ mem_roots hp0 |>.mp hx
+
+theorem roots_zero_iff_zero_or_isRoot_bot : p.roots = 0 ↔ p = 0 ∨ p.IsRoot = ⊥ := by
```
```suggestion
theorem roots_eq_zero_iff_eq_zero_or_isRoot_eq_bot : p.roots = 0 ↔ p = 0 ∨ p.IsRoot = ⊥ := by
```

erdOne commented on Mathlib/Algebra/Polynomial/Roots.lean:
```diff
@@ -581,6 +603,10 @@ theorem mem_rootSet_of_injective [CommRing S] {p : S[X]} [Algebra S R]
   classical
   exact Multiset.mem_toFinset.trans (mem_roots_map_of_injective h hp)
 
+theorem mem_rootSet'' [CommRing S] {p : S[X]} [Algebra S R] [FaithfulSMul S R] {x : R}
```
Can you add a docstring describing its difference with the unprimed versions? Thanks.

erdOne commented on Mathlib/Algebra/Polynomial/Roots.lean:
```diff
@@ -710,12 +736,26 @@ theorem Monic.isUnit_leadingCoeff_of_dvd {a p : R[X]} (hp : Monic p) (hap : a 
     IsUnit a.leadingCoeff :=
   isUnit_of_dvd_one (by simpa only [hp.leadingCoeff] using leadingCoeff_dvd_leadingCoeff hap)
 
+theorem Irreducible.roots_card_le_one (hirr : Irreducible p) : p.roots.card ≤ 1 := by
```
Is it intentional that this is in the `Polynomial.Irreducible` namespace? I'm not sure if this is a good choice.

erdOne commented on Mathlib/Algebra/Polynomial/Roots.lean:
```diff
@@ -710,12 +736,26 @@ theorem Monic.isUnit_leadingCoeff_of_dvd {a p : R[X]} (hp : Monic p) (hap : a 
+  exact (natDegree_eq_of_degree_eq_some <| degree_eq_one_of_irreducible_of_root hirr <|
+    isRoot_of_mem_roots hx).symm
+
+theorem Irreducible.roots_zero_of_natDegree_ne_one (hirr : Irreducible p) (hdeg : p.natDegree ≠ 1) :
```
```suggestion
theorem Irreducible.roots_eq_zero_of_natDegree_ne_one (hirr : Irreducible p) (hdeg : p.natDegree ≠ 1) :
```

erdOne commented on Mathlib/Algebra/Polynomial/Roots.lean:
```diff
@@ -784,6 +824,27 @@ theorem map_roots_le_of_injective [IsDomain A] [IsDomain B] (p : A[X]) {f : A 
   · simp only [hp0, roots_zero, Multiset.map_zero, Polynomial.map_zero, le_rfl]
   exact map_roots_le ((Polynomial.map_ne_zero_iff hf).mpr hp0)
 
+theorem map_roots_card_le_degree {A B : Type*} [Semiring A] [CommRing B] [IsDomain B]
```
```suggestion
theorem card_roots_map_le_degree {A B : Type*} [Semiring A] [CommRing B] [IsDomain B]
```
Because the LHS is `card (roots (map p f))` and the dot notation is merely notation. Same below.

erdOne commented on Mathlib/Algebra/Polynomial/Roots.lean:
```diff
@@ -784,6 +824,27 @@ theorem map_roots_le_of_injective [IsDomain A] [IsDomain B] (p : A[X]) {f : A 
+  · simp [hp0, hpm0, Polynomial.zero_le_degree_iff]
+  exact card_roots hpm0 |>.trans degree_map_le
+
+theorem map_roots_card_le_natDegree {A B : Type*} [Semiring A] [CommRing B] [IsDomain B]
```
```suggestion
theorem card_roots_map_le_natDegree {A B : Type*} [Semiring A] [CommRing B] [IsDomain B]
```

erdOne commented on Mathlib/Algebra/Polynomial/Roots.lean:
```diff
@@ -784,6 +824,27 @@ theorem map_roots_le_of_injective [IsDomain A] [IsDomain B] (p : A[X]) {f : A 
+    {f : A →+* B} (p : A[X]) : (p.map f).roots.card ≤ p.natDegree :=
+  card_roots' _ |>.trans natDegree_map_le
+
+theorem map_roots_filter_range_eq_roots_map [IsDomain A] [IsDomain B] {f : A →+* B}
```
```suggestion
theorem filter_roots_map_range_eq_map_roots [IsDomain A] [IsDomain B] {f : A →+* B}
```

erdOne commented on Mathlib/Algebra/Polynomial/Splits.lean:
```diff
@@ -585,21 +585,29 @@ theorem eval_derivative_div_eval_of_ne_zero_of_splits {p : K[X]} {x : K}
   rw [eval_derivative_eq_eval_mul_sum_of_splits h hx]
   exact mul_div_cancel_left₀ _ hx
 
+theorem prod_roots_mul_leadingCoeff_eq_coeff_zero_of_splits {P : K[X]}
```
```suggestion
theorem coeff_zero_eq_leadingCoeff_mul_prod_roots_of_splits {P : K[X]}
```
Or arguably these could go under the `Splits` namespace.

erdOne commented on Mathlib/Algebra/Polynomial/Splits.lean:
```diff
@@ -585,21 +585,29 @@ theorem eval_derivative_div_eval_of_ne_zero_of_splits {p : K[X]} {x : K}
+  convert prod_roots_mul_leadingCoeff_eq_coeff_zero_of_splits hP
+  simp [hmo]
+
+theorem sum_roots_mul_leadingCoeff_eq_nextCoeff_of_split {P : K[X]}
```
```suggestion
theorem nextCoeff_eq_neg_sum_roots_mul_leadingCoeff_of_splits {P : K[X]}
```

erdOne commented on Mathlib/FieldTheory/IsAlgClosed/AlgebraicClosure.lean:
```diff
@@ -200,6 +200,9 @@ instance {L : Type*} [Field k] [Field L] [Algebra k L] [Algebra.IsAlgebraic k L]
   isAlgebraic := .trans k L _
   isAlgClosed := inferInstance
 
+theorem aroots_card_eq_natDegree {p : k[X]} : (p.aroots <| AlgebraicClosure k).card = p.natDegree :=
```
Do we really want this lemma? Is `IsAlgClosed.aroots_card_eq_natDegree` not enough?

NotWearingPants replied:
This version doesn't require `Field` and `IsAlgClosed` and `FaithfulSMul` because those are known about `AlgebraicClosure`, but I deleted it

erdOne commented on Mathlib/FieldTheory/IsAlgClosed/Basic.lean:
```diff
@@ -108,25 +108,82 @@ theorem roots_eq_zero_iff [IsAlgClosed k] {p : k[X]} :
     rw [← mem_roots (ne_zero_of_degree_gt hd), h] at hz
     simp at hz
 
+theorem roots_eq_zero_iff_natDegree_zero [IsAlgClosed k] {p : k[X]} :
```
```suggestion
theorem roots_eq_zero_iff_natDegree_eq_zero [IsAlgClosed k] {p : k[X]} :
```

erdOne commented on Mathlib/FieldTheory/IsAlgClosed/Basic.lean:
```diff
@@ -108,25 +108,82 @@ theorem roots_eq_zero_iff [IsAlgClosed k] {p : k[X]} :
+    p.roots = 0 ↔ p.natDegree = 0 :=
+  roots_eq_zero_iff.trans eq_C_coeff_zero_iff_natDegree_eq_zero
+
+theorem roots_eq_zero_iff_degree_le_zero [IsAlgClosed k] {p : k[X]} : p.roots = 0 ↔ p.degree ≤ 0 :=
```
```suggestion
theorem roots_eq_zero_iff_degree_nonpos [IsAlgClosed k] {p : k[X]} : p.roots = 0 ↔ p.degree ≤ 0 :=
```

erdOne commented on Mathlib/FieldTheory/IsAlgClosed/Basic.lean:
```diff
@@ -108,25 +108,82 @@ theorem roots_eq_zero_iff [IsAlgClosed k] {p : k[X]} :
+theorem roots_eq_zero_iff_degree_le_zero [IsAlgClosed k] {p : k[X]} : p.roots = 0 ↔ p.degree ≤ 0 :=
+  roots_eq_zero_iff_natDegree_zero.trans natDegree_eq_zero_iff_degree_le_zero
+
+theorem roots_card_eq_natDegree [IsAlgClosed k] {p : k[X]} : p.roots.card = p.natDegree := by
```
```suggestion
theorem card_roots_eq_natDegree [IsAlgClosed k] {p : k[X]} : p.roots.card = p.natDegree := by
```

erdOne commented on Mathlib/FieldTheory/IsAlgClosed/Basic.lean:
```diff
@@ -108,25 +108,82 @@ theorem roots_eq_zero_iff [IsAlgClosed k] {p : k[X]} :
+  have ⟨_, _, hdeg, hroots⟩ := exists_prod_multiset_X_sub_C_mul p
+  simp [← hdeg, roots_eq_zero_iff_natDegree_zero.mp hroots]
+
+theorem map_roots_card_eq_natDegree_of_leadingCoeff_ne_zero {A B : Type*} [Semiring A] [Field B]
+    [IsAlgClosed B] (f : A →+* B) {p : A[X]} (hf : f p.leadingCoeff ≠ 0) :
+    (p.map f).roots.card = p.natDegree :=
+  natDegree_map_of_leadingCoeff_ne_zero _ hf ▸ roots_card_eq_natDegree
+
+theorem map_roots_card_eq_natDegree_of_isUnit_leadingCoeff {A B : Type*} [Semiring A] [Field B]
+    [IsAlgClosed B] (f : A →+* B) {p : A[X]} (h : IsUnit p.leadingCoeff) :
+    (p.map f).roots.card = p.natDegree :=
+  natDegree_map_of_isUnit_leadingCoeff f h ▸ roots_card_eq_natDegree
+
+theorem map_roots_card_eq_natDegree_of_injective {A B : Type*} [Semiring A] [Field B]
+    [IsAlgClosed B] (f : A →+* B) {p : A[X]} (hf : Function.Injective f) :
+    (p.map f).roots.card = p.natDegree :=
+  natDegree_map_eq_of_injective hf ▸ roots_card_eq_natDegree
+
+theorem map_roots_card_eq_natDegree_from_simpleRing {A B : Type*} [Ring A] [IsSimpleRing A]
+    [Field B] [IsAlgClosed B] (f : A →+* B) {p : A[X]} : (p.map f).roots.card = p.natDegree :=
+  natDegree_map_from_simpleRing f p ▸ roots_card_eq_natDegree
+
+theorem aroots_card_eq_natDegree_of_leadingCoeff_ne_zero {A B : Type*} [CommRing A] [Field B]
+    [IsAlgClosed B] [Algebra A B] {p : A[X]} (hf : algebraMap A B p.leadingCoeff ≠ 0) :
+    (p.aroots B).card = p.natDegree :=
+  map_roots_card_eq_natDegree_of_leadingCoeff_ne_zero _ hf
+
+theorem aroots_card_eq_natDegree_of_isUnit_leadingCoeff {A B : Type*} [CommRing A] [Field B]
+    [IsAlgClosed B] [Algebra A B] {p : A[X]} (h : IsUnit p.leadingCoeff) :
+    (p.aroots B).card = p.natDegree :=
+  map_roots_card_eq_natDegree_of_isUnit_leadingCoeff _ h
+
+theorem aroots_card_eq_natDegree {A B : Type*} [CommRing A] [Field B] [IsAlgClosed B] [Algebra A B]
+    [FaithfulSMul A B] {p : A[X]} : (p.aroots B).card = p.natDegree :=
+  map_roots_card_eq_natDegree_of_injective _ <| FaithfulSMul.algebraMap_injective _ _
```
The names of these are wrong too.

erdOne commented on Mathlib/FieldTheory/IsAlgClosed/Basic.lean:
```diff
@@ -108,25 +108,82 @@ theorem roots_eq_zero_iff [IsAlgClosed k] {p : k[X]} :
-  exists_eval₂_eq_zero (algebraMap R k) p hp
+  exists_eval₂_eq_zero _ p hp
+
+theorem exists_aeval_eq_zero' {R : Type*} [CommSemiring R] [IsAlgClosed k] [Algebra R k]
```
Can you add a docstring describing the difference of this with the unprimed one? Thanks.

erdOne reviewed (commented):
Mostly nitpicks about names

SnirBroshi committed:
feat(Algebra/Polynomial/Eval/Coeff): zero is a root iff no constant coefficient

SnirBroshi committed:
feat(Algebra/Polynomial/Degree/Lemmas): `nextCoeff` distributes over multiplying by a constant

SnirBroshi committed:
feat(Algebra/Polynomial/Eval/Defs): applying a function on an evaluation distributes into all the coefficients and the variable

SnirBroshi committed:
feat(Algebra/Polynomial/Basic): `C ofNat = ofNat`

SnirBroshi committed:
feat(Algebra/Polynomial/Div): `p ∣ q → rootMultiplicity x p ≤ rootMultiplicity x q`

SnirBroshi committed:
feat(Algebra/Polynomial/Div): irreducible polynomials have at most one root

SnirBroshi committed:
feat(Algebra/Polynomial/Roots): `roots = 0 ↔ IsRoot = ⊥`

SnirBroshi committed:
feat(Algebra/Polynomial/Roots): irreducible polynomials have at most one root

SnirBroshi committed:
feat(Algebra/Polynomial/Roots): the amount of roots after a map is still bound by the degree

SnirBroshi committed:
feat(Algebra/Polynomial/Roots): any new roots after a map are not in the range of the map

SnirBroshi committed:
feat(Algebra/Polynomial/Splits): sum and product of the roots show up as coefficients even if not monic

SnirBroshi committed:
feat(Algebra/Polynomial/Eval/Degree): add `nextCoeff_map_of_leadingCoeff_ne_zero` and golf

SnirBroshi committed:
feat(Algebra/Polynomial/Degree/Lemmas): degrees are preserved when the leading coeff is a unit

SnirBroshi committed:
feat(Algebra/Polynomial/Degree/Lemmas): degrees are preserved under injective maps

SnirBroshi committed:
feat(Algebra/Polynomial/Degree/Lemmas): degrees are preserved under maps from simple rings

SnirBroshi committed:
chore(Algebra/Polynomial/FieldDivision): generalize DivisionRing to `IsSimpleRing` and simplify proofs using existing theorems

SnirBroshi committed:
feat(Algebra/Polynomial/AlgebraMap): add `isRoot_of_aeval_algebraMap_eq_zero'` that requires `FaithfulSMul` instead of an injective `algebraMap`

SnirBroshi committed:
feat(RingTheory/SimpleRing/Basic): add `FaithfulSMul` instance for algebras over simple rings

SnirBroshi committed:
feat(FieldTheory/IsAlgClosed/Basic): generalize `exists_eval₂_eq_zero` from `DivisionRing` to `IsSimpleRing`

SnirBroshi committed:
feat(FieldTheory/IsAlgClosed/Basic): generalize `exists_aeval_eq_zero` from `Field` to `CommSemiring` + `FaithfulSMul`

SnirBroshi committed:
feat(Algebra/Polynomial/Roots): associated polynomials have the same roots

SnirBroshi committed:
feat(FieldTheory/IsAlgClosed/Basic): add theorems relating roots and divisibility

SnirBroshi committed:
chore(Algebra/Polynomial/Roots): explain the differences between `mem_rootSet` and `mem_rootSet'`

SnirBroshi committed:
feat(FieldTheory/IsAlgClosed/Basic): only constant polynomial have no roots

SnirBroshi committed:
feat(FieldTheory/IsAlgClosed/Basic): the amount of roots equals the degree

SnirBroshi committed:
feat(FieldTheory/IsAlgClosed/Basic): many conditions for the amount of roots after a map to equal the degree before the map

SnirBroshi committed:
feat(FieldTheory/IsSepClosed): add `exists_eval₂_eq_zero_of_injective`

SnirBroshi committed:
feat(FieldTheory/IsSepClosed): generalize `exists_eval₂_eq_zero` from `Field` to `CommRing` + `IsSimpleRing`

SnirBroshi committed:
feat(FieldTheory/IsSepClosed): generalize `exists_aeval_eq_zero` from `Field` to `CommSemiring` + `FaithfulSMul`

ocfnash reviewed (commented):
Thanks, there are numerous useful lemmas here!

However this is very difficult to review in its current state. One of the reasons is that you are changing our imports and even proposing dropping an `assert_not_exists` in parallel with adding a large number of minor new results.

I think to progress here you should:

- split this PR up into several pieces
- attempt to eliminate or reduce the import increases
- consider means to obviate removing the `assert_not_exists`

Perhaps you could start by gathering those results which do nothing controversial into a PR of their own.

SnirBroshi committed:
Merge remote-tracking branch 'origin/master' into feature/polynomial-lemmas

SnirBroshi committed:
chore(Algebra/Polynomial/Div): move `aeval_ne_zero_of_irredicble_natDegree_ne_one` to `RingDivision` to avoid an import

SnirBroshi committed:
chore(Algebra/Polynomial/Degree/Lemmas): move simple-ring theorems to `FieldDivision` to avoid the import and deduplicate

SnirBroshi committed:
Revert "feat(RingTheory/SimpleRing/Basic): add `FaithfulSMul` instance for algebras over simple rings"

This reverts commit feefe2f3e34db4f8a2c15836bba72a42dec369da.

PR #29356 added the same instance to a new module, `Mathlib.Algebra.Algebra.IsSimpleRing`

SnirBroshi committed:
chore(Algebra/FieldTheory/IsAlgClosed/Basic): ungeneralize `exists_aeval_eq_zero` from `CommSemiring` + `FaithfulSMul` to `CommRing` + `IsSimpleRing`

For `FaithfulSMul` to be a generalization we need to import `Mathlib.Algebra.Algebra.IsSimpleRing` which contains an instance of `FaithfulSMul` given `IsSimpleRing`

SnirBroshi commented on Mathlib/RingTheory/SimpleRing/Basic.lean:
```diff
@@ -22,8 +23,6 @@ A ring `R` is **simple** if it has only two two-sided ideals, namely `⊥` and `
 
 -/
 
-assert_not_exists Finset
```
Reverted, the same instance was added in #29356 

> Please do not resolve conversations like this without agreement from the person who started it.

Sorry, I was given the impression from another reviewer that I must resolve everything I've responded to.

Now that this is reverted I think I can resolve (without agreement from the person who started it) as it objectively no longer matters.

SnirBroshi commented:
> Thanks, there are numerous useful lemmas here!
> 
> However this is very difficult to review in its current state. One of the reasons is that you are changing our imports and even proposing dropping an `assert_not_exists` in parallel with adding a large number of minor new results.
> 
> I think to progress here you should:
> 
> * split this PR up into several pieces
> * attempt to eliminate or reduce the import increases
> * consider means to obviate removing the `assert_not_exists`
> 
> Perhaps you could start by gathering those results which do nothing controversial into a PR of their own.

- No more import changes
- No more instances
- No more removing `assert_not_exists`

Hopefully nothing controversial remains.

SnirBroshi committed:
Merge remote-tracking branch 'origin/master' into feature/polynomial-lemmas

SnirBroshi committed:
chore(Algebra/Polynomial): small golfs

ocfnash commented on Mathlib/Algebra/Polynomial/Degree/Lemmas.lean:
```diff
@@ -271,6 +271,39 @@ theorem natDegree_map_eq_iff {f : R →+* S} {p : Polynomial R} :
+
+theorem leadingCoeff_map_of_injective {f : R →+* S} (hf : Function.Injective f)
+    (p : Polynomial R) : (p.map f).leadingCoeff = f p.leadingCoeff := by
+  grind [leadingCoeff, natDegree_map_eq_of_injective, coeff_map]
```
```suggestion
  simp only [leadingCoeff, natDegree_map_eq_of_injective hf, coeff_map]
```

ocfnash commented on Mathlib/Algebra/Polynomial/Div.lean:
```diff
@@ -743,6 +748,19 @@ lemma degree_eq_one_of_irreducible_of_root (hi : Irreducible p) {x : R} (hx : Is
+    (hi : Irreducible p) (hdeg : p.natDegree ≠ 1) {x : R} : ¬p.IsRoot x :=
+  fun hr ↦ hdeg <| natDegree_eq_of_degree_eq_some <| degree_eq_one_of_irreducible_of_root hi hr
+
+lemma isRoot_eq_bot_of_irreducible_natDegree_ne_one
```
```suggestion
lemma _root_.Irreducible.isRoot_eq_bot_of_natDegree_ne_one
```

ocfnash commented on Mathlib/Algebra/Polynomial/Roots.lean:
```diff
@@ -111,6 +111,17 @@ theorem ne_zero_of_mem_roots (h : a ∈ p.roots) : p ≠ 0 :=
+  by_cases hp0 : p = 0
+  · exact ⟨fun _ ↦ Or.inl hp0, fun _ ↦ hp0 ▸ roots_zero⟩
+  refine ⟨fun h ↦ Or.inr <| roots_eq_zero_iff_isRoot_eq_bot hp0 |>.mp h, fun h ↦ Or.elim h ?_ ?_⟩
+  exacts [fun h ↦ False.elim <| hp0 h, fun h ↦ roots_eq_zero_iff_isRoot_eq_bot hp0 |>.mpr h]
```
```suggestion
  rcases eq_or_ne p 0 with rfl | hp0; · simp
  simp [roots_eq_zero_iff_isRoot_eq_bot hp0, hp0]
```

ocfnash commented on Mathlib/Algebra/Polynomial/Roots.lean:
```diff
@@ -189,6 +200,11 @@ theorem roots_C_mul (p : R[X]) (ha : a ≠ 0) : (C a * p).roots = p.roots := by
     simp only [roots_mul, *, Ne, mul_eq_zero, C_eq_zero, or_self_iff, not_false_iff, roots_C,
       zero_add, mul_zero]
 
+theorem roots_congr {p q : R[X]} (h : Associated p q) : p.roots = q.roots := by
```
```suggestion
theorem _root_.Associated.roots_eq {p q : R[X]} (h : Associated p q) : p.roots = q.roots := by
```

ocfnash reviewed (approved):
Thanks!

Please apply the suggestions and then feel free to merge.

bors d+

SnirBroshi committed:
chore(Algebra/Polynomial/Eval): ungolf because `grind` is still unstable

SnirBroshi committed:
chore(Algebra/Polynomial/AlgebraMap): remove `isRoot_of_aeval_algebraMap_eq_zero` in favor of the primed version

SnirBroshi committed:
chore(Algebra/Polynomial/Div): rename `not_isRoot_of_irreducible_natDegree_ne_one` -> `Irreducible.not_isRoot_of_natDegree_ne_one`

SnirBroshi committed:
chore(Algebra/Polynomial/FieldDivision): rename `nextCoeff_map_eq_from_simpleRing` -> `nextCoeff_map_eq`

SnirBroshi committed:
chore(Algebra/Polynomial/FieldDivision): tag `monic_map_iff` with simp

SnirBroshi committed:
chore(Algebra/Polynomial/Roots): rename `roots_congr` -> `Associated.roots_eq`

SnirBroshi committed:
Apply suggestions from code review

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

SnirBroshi committed:
Merge branch 'master' into feature/polynomial-lemmas
